wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/EnvironmentParser.java

```diff
@@ -78,36 +85,79 @@ public CheServicesEnvironmentImpl parse(Environment environment) throws IllegalA
                       "Recipe of environment must contain location or content");
 
-        CheServicesEnvironmentImpl composeEnvironment;
+        CheServicesEnvironmentImpl cheServicesEnvironment;
         String envType = environment.getRecipe().getType();
         switch (envType) {
             case "compose":
-                composeEnvironment = parseCompose(environment.getRecipe());
+                cheServicesEnvironment = parseCompose(environment.getRecipe());
                 break;
             case "dockerimage":
             case "dockerfile":
-                composeEnvironment = parseDocker(environment);
+                cheServicesEnvironment = parseDocker(environment);
                 break;
             default:
-                throw new IllegalArgumentException("Environment type " + envType + " is not supported");
+                throw new IllegalArgumentException(format("Environment type '%s' is not supported. " +
+                                                          "Supported environment types: %s",
+                                                          envType,
+                                                          Joiner.on(", ").join(types)));
         }
 
-        composeEnvironment.getServices().forEach((name, service) -> {
+        cheServicesEnvironment.getServices().forEach((name, service) -> {
             ExtendedMachine extendedMachine = environment.getMachines().get(name);
-            if (extendedMachine != null &&
-                extendedMachine.getAttributes() != null &&
-                extendedMachine.getAttributes().containsKey("memoryLimitBytes")) {
-
-                try {
-                    service.setMemLimit(Long.parseLong(extendedMachine.getAttributes().get("memoryLimitBytes")));
-                } catch (NumberFormatException e) {
-                    throw new IllegalArgumentException(
-                            format("Value of attribute 'memoryLimitBytes' of machine '%s' is illegal", name));
-                }
+            if (extendedMachine != null) {
+                normalizeMachine(name, service, extendedMachine);
             }
         });
 
-        return composeEnvironment;
+        return cheServicesEnvironment;
+    }
+
+    private void normalizeMachine(String name, CheServiceImpl service, ExtendedMachine extendedMachine) {
+        if (extendedMachine.getAttributes() != null &&
+            extendedMachine.getAttributes().containsKey("memoryLimitBytes")) {
+
+            try {
+                service.setMemLimit(Long.parseLong(extendedMachine.getAttributes().get("memoryLimitBytes")));
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException(
+                        format("Value of attribute 'memoryLimitBytes' of machine '%s' is illegal", name));
+            }
+        }
+        if (service.getExpose() != null) {
```

Comment:
the same to https://github.com/eclipse/che/pull/2599/files#r80632123
CheServiceImpl#getExpose returns new ArrayList<String>() instead of null https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/model/CheServiceImpl.java#L285-290